cwrapper/CArachneWrapper.h

```diff
@@ -0,0 +1,60 @@
+/* Copyright (c) 2017 Stanford University
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR(S) DISCLAIM ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL AUTHORS BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#ifndef CARACHNEWRAPPER_H_
+#define CARACHNEWRAPPER_H_
+
+#include <inttypes.h>
+#include <stdio.h>
+#include <sys/types.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * \addtogroup api Arachne C Wrapper
+ * In order to use Arachne in applications written in C, we provide C wrappers
+ * for Arachne public API functions
+ * @{
+ */
+/**
+ * This structure is the C wrapper for Arachne::ThreadContext
+ * and can only be used as pointers.
+ */
+struct CArachneThreadContext;
+typedef struct CArachneThreadContext CArachneThreadContext;
+/**
+ * This structure is the C wrapper for Arachne::ThreadId.
+ */
+struct CArachneThreadId {
```

Comment:
What would be the cost to avoiding these "mirror" structures? Does it mean that we can only expose pointers in the C interface?

Reply:
Yes we can only use pointers in the C interface. We may also need to change the C++ interface to only return pointers. 

Because if we return an object (e.g., ThreadId) from a C++ function (e.g., createThread), it will unlikely to directly use it in C or convert it into a pointer. That's why I need to work around ThreadId class by mirroring to CArachneThreadId.

Reply:
Probably we can keep this mirror structure now? We may not want to change C++ interface.

Reply:
Agreed.